### PR TITLE
Add upper bound check to all

### PR DESCRIPTION
**Pull request by jappeace**
reacration of https://github.com/haskell/cabal/pull/8339

fixes https://github.com/haskell/cabal/issues/8291
presumably this will make it nag at anyone for forgetting
to add upper bounds to their packages.

I tested this by running cabal check on yesod-auth-simple (which has no upper bounds)
then added said upper bounds with cabal gen-bounds, and now it works.

I also added a test public-multilib-3.cabal and public-multilib-3.check which expects the output.
Furthermore an upper bound was added to public-multilib-2.cabal to assert it's 
not emmiting a warning.

---
Please include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#conventions).
* [x] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [x] The documentation has been updated, if necessary.

### CODE REVIEW

**Comment by Mikolaj** [comment]
@mergify refresh

**Comment by Mikolaj** [comment]
Perhaps mergify is not mollified after the old PR has been closed. We'll see. :)

File: Cabal/src/Distribution/PackageDescription/Check.hs
Comment (ffaf1) [inline Cabal/src/Distribution/PackageDescription/Check.hs:678]: is parsonmatt better than — say — https://pvp.haskell.org/ ?

Comment (jappeace) [reply Cabal/src/Distribution/PackageDescription/Check.hs]: I changed it to pvp, I originally did this post because it explains well why upperbounds are required, the pvp page only says you must do this, no reasoning, which is unsatisfying to me.

**Comment by Mikolaj** [comment]
CI is acting up, so let me rebase to include the our CI workaround.

**Comment by Mikolaj** [comment]
@mergify rebase

File: Cabal/src/Distribution/PackageDescription/Check.hs
Comment (andreasabel) [inline Cabal/src/Distribution/PackageDescription/Check.hs:676]: "with with" --> "e.g. using"
`cabal gen-bounds` cannot add these bounds directly, but can print out suggestions.

File: Cabal/src/Distribution/PackageDescription/Check.hs
Comment (andreasabel) [inline Cabal/src/Distribution/PackageDescription/Check.hs:1830]: Spurious parentheses.

File: Cabal/src/Distribution/PackageDescription/Check.hs
Comment (andreasabel) [inline Cabal/src/Distribution/PackageDescription/Check.hs:2463]: I think this would be easier to read with a list comprehension:
```haskell
  Map.fromListWIth intersectVersionRanges
    [ (pname, vr) 
    | Dependency pname vr _ <- selectDependencies pkgs'
    , pname /= self
    ]
```

File: Cabal/src/Distribution/PackageDescription/Check.hs
Comment (andreasabel) [inline Cabal/src/Distribution/PackageDescription/Check.hs:2413]: Section preferred: `((setupDepends =<<) . maybeToList . setupBuildInfo)`
Or, even better: `maybe [] setupDepends . setupBuildInfo`
Yeah, why turn `Maybe` into a 0-1 list and then use `concatMap`?

(Thought about: `=<<`:
That said, I am not a great fan of using `=<<` for lists...
The original list comprehension looked much more readable than the reformulation with `=<<` and `do`.
(But I acknowledge you wanted to factor out common code.))

**Review by andreasabel — CHANGES_REQUESTED** [review]
Thanks for the PR!

Got some style critique...

File: Cabal/src/Distribution/PackageDescription/Check.hs
Comment (jappeace) [reply Cabal/src/Distribution/PackageDescription/Check.hs]: I changed that into "please add them using `cabal gen-bounds` for suggestions."

File: Cabal/src/Distribution/PackageDescription/Check.hs
Comment (jappeace) [reply Cabal/src/Distribution/PackageDescription/Check.hs]: I removed these

File: Cabal/src/Distribution/PackageDescription/Check.hs
Comment (jappeace) [reply Cabal/src/Distribution/PackageDescription/Check.hs]: I changed thsi to the list comprehension like suggested.

File: Cabal/src/Distribution/PackageDescription/Check.hs
Comment (jappeace) [reply Cabal/src/Distribution/PackageDescription/Check.hs]: I used `foldMap setupDepends . setupBuildInfo`

I only touched this originally because I was looking at the wrong function somhow, and then I realized these were incredibly similar so I factored it out :sweat_smile: 
I don't particularly favor bind or list comprehensions in any case,
I just hadn't see list comprehensions used that much before in haskell code, so part of my understanding process was changing it into it's bind based representation.
>>= on lists can go really unexpected because it does cartesian products by default instead of pointwise products like ziplist does

Comment (andreasabel) [reply Cabal/src/Distribution/PackageDescription/Check.hs]: Thanks for fixing!

File: Cabal/src/Distribution/PackageDescription/Check.hs
Comment (jappeace) [reply Cabal/src/Distribution/PackageDescription/Check.hs]: I replaced this with a list comprehension.

**Comment by jappeace** [comment]
If nobody has anything else to review, please merge?

File: Cabal/src/Distribution/PackageDescription/Check.hs
Comment (ffaf1) [inline Cabal/src/Distribution/PackageDescription/Check.hs:229]: Is there a need to leave `BaseNoUpperBounds` with your patch?

File: cabal-install/tests/UnitTests/Distribution/Solver/Modular/DSL.hs
Comment (ffaf1) [inline cabal-install/tests/UnitTests/Distribution/Solver/Modular/DSL.hs:481]: Could this be expressed with an `&&` inside the filter?
Why do we need to filter `MissingUpperBound` here?

File: Cabal/src/Distribution/PackageDescription/Check.hs
Comment (jappeace) [reply Cabal/src/Distribution/PackageDescription/Check.hs:229]: It get's a different warning level, `PackageDistInexcusable` for base not having upper bounds, and `PackageDistSuspiciousWarn` for other dependencies.

I hoped that it'd just nag at the user for not having an upper bound on other dependencies then base.

File: cabal-install/tests/UnitTests/Distribution/Solver/Modular/DSL.hs
Comment (jappeace) [reply cabal-install/tests/UnitTests/Distribution/Solver/Modular/DSL.hs]: A whole host of tests started failing because the generated package crashed on the check call in the generator, for example:
```
      show conflicts from final conflict set after exhaustive search:                                                                  FAIL
        Exception: invalid GenericPackageDescription for package A-1.0.0: [These packages miss upper bounds 'B','F' please add them using `cabal gen-bounds` for suggestions. For more information see:  https://pvp.haskell.org/]
        CallStack (from HasCallStack):
          error, called at tests/UnitTests/Distribution/Solver/Modular/DSL.hs:484:13 in main:UnitTests.Distribution.Solver.Modular.DSL
        Use -p '/show conflicts from final conflict set after exhaustive search/' to rerun this test only.
      show first conflicts after inexhaustive search:                                                                                  FAIL
        Exception: invalid GenericPackageDescription for package A-1.0.0: [These packages miss upper bounds 'B','F' please add them using `cabal gen-bounds` for suggestions. For more information see:  https://pvp.haskell.org/]
        CallStack (from HasCallStack):
```
filtering out the missing upper bound warning here makes the test work as intended.

Comment (Mikolaj) [reply cabal-install/tests/UnitTests/Distribution/Solver/Modular/DSL.hs]: Makes sense.

**Comment by Mikolaj** [comment]
@andreasabel: could you re-review?

File: Cabal-tests/tests/ParserTests/regressions/public-multilib-3.check
Comment (andreasabel) [inline Cabal-tests/tests/ParserTests/regressions/public-multilib-3.check:3]: Being a bit nit-picky, but this sentence lacks punctuation.
```
These packages miss upper bounds:
- somelib
Please add them, using ...
```
I am suggesting also to print the list via bullet points, makes it easier to read if there are many.
Yeah, why not make a test case (or modify this one) with a long list of packages that lack a bound?

**Review by andreasabel — CHANGES_REQUESTED** [review]
Thanks for the update.

I have suggestions for the error message formatting.

File: cabal-install/tests/UnitTests/Distribution/Solver/Modular/DSL.hs
Comment (ffaf1) [reply cabal-install/tests/UnitTests/Distribution/Solver/Modular/DSL.hs]: I see, but then why not just one filter instead of two? `\x -> (not . isSomething) x && (not . isSomethingElse x)`

Comment (Mikolaj) [reply cabal-install/tests/UnitTests/Distribution/Solver/Modular/DSL.hs]: That's outdated code. This is just one filter now.

File: Cabal/src/Distribution/PackageDescription/Check.hs
Comment (gbaz) [reply Cabal/src/Distribution/PackageDescription/Check.hs]: the pvp faq has reasoning on upper bounds https://pvp.haskell.org/faq/#upper-bounds

File: Cabal-tests/tests/ParserTests/regressions/public-multilib-3.check
Comment (jappeace) [reply Cabal-tests/tests/ParserTests/regressions/public-multilib-3.check]: I changed this like you suggested

**Comment by jappeace** [comment]
If nobody has anything else to review, please merge?

File: Cabal/src/Distribution/PackageDescription/Check.hs
Comment (ffaf1) [inline Cabal/src/Distribution/PackageDescription/Check.hs:1826]: This is minor but instead of `if` is there a reason why you cannot do something like: `map (packageDistSuspiciousWarn . MissingUpperBounds) other : …`?

**Comment by ffaf1** [comment]
I have added a small comment, also there are some failures in CI, do you need help with those (doctest, etc)?

File: cabal-install/tests/UnitTests/Distribution/Solver/Modular/DSL.hs
Comment (grayjay) [inline cabal-install/tests/UnitTests/Distribution/Solver/Modular/DSL.hs:481]: Please update the comment above this filter, since the missing bounds check is unrelated to unknown extensions and languages.

File: Cabal-tests/tests/ParserTests/regressions/public-multilib-3.check
Comment (ulysses4ever) [reply Cabal-tests/tests/ParserTests/regressions/public-multilib-3.check]: @andreasabel are you happy with it now? Could you re-review again?

File: cabal-install/tests/UnitTests/Distribution/Solver/Modular/DSL.hs
Comment (jappeace) [reply cabal-install/tests/UnitTests/Distribution/Solver/Modular/DSL.hs]: I updated this comment

**Comment by mouse07410** [comment]
Just to mention that often I do not "forget" but deliberately omit upper bound. Enforcing it sometimes makes sense, and sometimes absolutely doesn't.

**Comment by jappeace** [comment]
@mouse07410 this should be implemented as just a warning. It shouldn't block (unless I did it wrong).

**Comment by jappeace** [comment]
@ffaf1 yes I've no idea why the doctests are failing, it doesn't seem to be related to my changes

File: Cabal/src/Distribution/PackageDescription/Check.hs
Comment (jappeace) [reply Cabal/src/Distribution/PackageDescription/Check.hs]: Yes! That works, thanks!

**Comment by ffaf1** [comment]
Seems a connection timed out, not your fault. There are compilation erros in your patch tho, fix those and you should be fine.

**Comment by Mikolaj** [comment]
Actually, CI was corrupted at that time and it's now fixed, so let me rebase and see if CI passes now.

**Comment by Mikolaj** [comment]
@mergify rebase

**Comment by Mikolaj** [comment]
CI passes! What now?

File: Cabal-tests/tests/ParserTests/regressions/all-upper-bound.check
Comment (ffaf1) [inline Cabal-tests/tests/ParserTests/regressions/all-upper-bound.check:1]: Nitpick: Add a `maintainer` and `description` in the cabal file. This way if description error changes there is one less thing to modify in the future.

**Review by ffaf1 — APPROVED** [review]
Other the few things I highlighted, I approve.

File: Cabal/src/Distribution/PackageDescription/Check.hs
Comment (jappeace) [reply Cabal/src/Distribution/PackageDescription/Check.hs:1829]: It's needed because otherwise this will always emit an error, even if others is empty, others and baseError are different types!

Comment (jappeace) [reply Cabal/src/Distribution/PackageDescription/Check.hs:1829]: we could, but that's how I had it before I got comments on this...
The advantage of this approach is that we can pack the warning per package and provide a nice listing to the user.

File: Cabal-tests/tests/ParserTests/regressions/all-upper-bound.check
Comment (jappeace) [reply Cabal-tests/tests/ParserTests/regressions/all-upper-bound.check]: I added the package and maintainer fields to the cabal file, thanks.

**Review by Kleidukos — APPROVED** [review]
Nice.

**Comment by Mikolaj** [comment]
@andreasabel seems to be busy, so let me dimiss his blockage. Anybody else objects to merging? @jappeace: anything else you'd lke to address or improve before the merge? If that's it, I'm going to set label squash+merge_me (unless you prefer no squash) RSN.

**Comment by Mikolaj** [comment]
@mergify rebase

**Comment by Mikolaj** [comment]
@jappeace: I'm afraid a test fails after rebasing, probably because it was added independently and the captured output doesn't contain the bounds warnings. Would you amend it?

**Comment by ulysses4ever** [comment]
I don’t want to argue one way or another, but I feel like some people will get upset about this change: after all, PVP doesn’t have universal support (in this part, at least), and there are Hackage authors who don’t put upper bounds purposefully. Given that it’s a mere warning (is it, btw?), I guess, they will survive. Still, I think, it may be nicer to at least supply a flag that turns the check off. Future work, perhaps…

Also, I’d hope for a more coherent explanation of the change in the **commit message**. Preferably, an explanation from the end user’s perspective.

**Comment by mouse07410** [comment]
I am totally with those who do not put upper bounds in purposefully.

So, add long as this PR results _**only**_ in warning in such cases, I agree to live with it, though I personally find the idea silly.

**Comment by ffaf1** [comment]
[In light of this](https://github.com/haskell/cabal/issues/8291#issuecomment-1185689545), I still believe a warning `PackageDistSuspiciousWarn` is a sensible thing to print. Third party tools (and `cabal-install`) can then filter out what they don’t fancy, butt first thing first!

**Comment by Mikolaj** [comment]
If this is controversial, we don't have to release it in 3.10 yet, but let's merge so that there's something to argue about and test in the field. It's bad style to let a PR slip. Let it go up in flames if unavoidable.

@ulysses4ever: if we are squash-merging, then indeed a good single commit message would be required instead of the default list of squashed commits. Is the changelog entry good enough? If so, perhaps copy it? If not, amend it?

@ffaf1: could you open a ticket about PackageDistSuspiciousWarn and/or a flag? It's fine if it refers to a yet unmerged PR.

**Comment by ulysses4ever** [comment]
As I said, I'm not against merging it, so, by all means…

Squash is the next level: presently, even the first commit message (which I usually expect to have some explanation no matter what) is kinda dense. Copying the changelog to the first commit message is an option (if the changelog has a decent explanation).

By the way, Mergify allows several approaches to forming the commit message on squash. We use the default one when it concatenates all messages. Another option is to take the first commit's message. I was wondering if the latter would be a better default. But this is off-topic here, perhaps...

**Comment by ffaf1** [comment]
`PackageDistSuspiciousWarn` is what will happen with this patch, if tools fancy it, they can filter it. I will open a ticket for a possible flag. (done!)

**Comment by mouse07410** [comment]
I have one more problem with this PR. Many packages, including those on Hackage, and projects set `Werror`, treating warnings as errors. Even though it's inconvenient but possible to change this setting in local projects, it's totally impractical to change it on others' packages you install from Hackage. This can break a ton of currently-working independent packages. 

Thus, I insist that if this PR is merged, it must output this message as **Info**, not as **Warning**.

**Comment by ffaf1** [comment]
This is a `cabal check` patch @mouse07410, it does not touch compilation stuff.

**Comment by mouse07410** [comment]
> This is a `cabal check` patch @mouse07410, it does not touch compilation stuff.

Thank you! Does it mean that it can be used only on local projects?

P.S. I'm still worried about adding a warning-level message, as I can't see all the consequences. At least, your post shows that they are limited...

**Comment by fgaz** [comment]
@mouse07410 There are no consequences except for the warning message. `cabal check` will still exit with code 0 and Hackage will still accept the package. See https://github.com/haskell/cabal/blob/1003ed665e2d1073b0ded10c2dcbe5bdd6bb18f9/Cabal/src/Distribution/PackageDescription/Check.hs#L790-L792

**Comment by ffaf1** [comment]
@jappeace could you please rebase it? We are so close to merging this.

**Comment by jappeace** [comment]
I shall do so now. :grin:

**Comment by ffaf1** [comment]
well there you have it, `description` should be longer than `synopsis` in your test. Patch it and it should be good!

**Comment by ffaf1** [comment]
Congrats @jappeace

**Comment by Kleidukos** [comment]
Congratulations @jappeace! The path was tortuous but you made it! :)

**Comment by gbaz** [comment]
@emilypi has pointed out that this checks test suites as well as exposed libs and executables. Arguably it should not do so (especially as gen-bounds does not generate bounds for test stanzas)

**Comment by ffaf1** [comment]
#8427 has a slightly more appropriate warning message:

```
Warning: On test suite 'test', these packages miss upper bounds:
- text
Please add them. More informations at https://pvp.haskell.org/
```

thanks to a suggestion by @andreasabel (indication of target, no mention of `gen-bounds`).

[PVP](https://pvp.haskell.org/#dependencies-in-cabal) does not seem to make differences between targets.

But the more useful question is: what policy does Hackage follow *in practice*? In the end this PR stems from a Hackage/cabal UX problem highlighted in #8291, `check` should mimic Hackage requirements.

**Comment by emilypi** [comment]
Tbqh if we're making it a requirement of Hackage that we have bounds on test targets, that's fine. I'd rather see `gen-bounds` do the proper thing (generate bounds for specified targets) rather than change the message. But yes, PVP also doesn't cover semantic relevance beyond bounds, and Hackage is deviating from PVP by considering tests to be a relevant part of the sdist. I'm actually on board with this. Tests and benchmarks are relevant to the spec for what constitutes correctness in many cases (e.g. cryptography).

**Comment by gbaz** [comment]
"what policy does Hackage follow in practice" -- hackage actually makes use of cabal check! The ticket discussion was misleading here. There is a _purely textual_ set of guidelines for packages which asks for but does not mandate compliance with the PVP. But in terms of the checks hackage runs on upload, those are fundamentally the checks from cabal _already_.

**Comment by gbaz** [comment]
I have always understood PVP to apply to libraries and executables, not internal targets, and this has always been the understanding of hackage trustees among others. So I believe the correct thing to do here is to remove upper bounds checks in cabal on test stanzas and benchmark stanzas. They are not considered mandated by the pvp it has never been hackage trustee policy to mandate them, and I think it was a purely accidental change in that regard made by this PR.

**Pull request merged**
